watcher/local/watcher.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/watcher"
+)
+
+type (
+	// Watcher implements a local watcher.
+	Watcher struct {
+		rs channel.RegisterSubscriber
+		*registry
+	}
+
+	// Config represents the configuration for initializing a local watcher.
+	Config struct {
+		RegisterSubscriber channel.RegisterSubscriber
+	}
+
+	ch struct {
+		id     channel.ID
+		params *channel.Params
+		parent *ch
+
+		// subChsAccess mutex is used for thread-safe access of a parent
+		// channel and all of its sub-channels. For example, while adding new
+		// sub-channels to a ledger channel or while registering dispute for
+		// the ledger channel and all its children.
+		subChsAccess sync.Mutex
+	}
+)
+
+// NewWatcher initializes a local watcher.
+//
+// Local watcher implements pub-sub interfaces over "go channels".
+func NewWatcher(cfg Config) (*Watcher, error) {
+	w := &Watcher{
+		rs:       cfg.RegisterSubscriber,
+		registry: newRegistry(),
+	}
+	return w, nil
+}
+
+// StartWatchingLedgerChannel starts watching for a ledger channel.
+func (w *Watcher) StartWatchingLedgerChannel(
+	ctx context.Context,
+	signedState channel.SignedState,
+) (watcher.StatesPub, watcher.AdjudicatorSub, error) {
+	return w.startWatching(ctx, nil, signedState)
+}
+
+// StartWatchingSubChannel starts watching for a sub-channel or virtual channel.
```

Comment:
Do you document somewhere that only 1 layer of subchannels is currently supported?

Reply:
I had not documented it earlier. Updated in 35d1578.